sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala

```diff
@@ -598,6 +598,37 @@ abstract class SQLQuerySuiteBase extends QueryTest with SQLTestUtils with TestHi
     }
   }
 
+  test("SPARK-28551: CTAS Hive Table should be with non-existent or empty location") {
+    def executeCTASWithNonEmptyLocation(tempLocation: String) {
+      sql(s"CREATE TABLE ctas1(id string) stored as rcfile LOCATION '$tempLocation/ctas1'")
+      sql("INSERT INTO TABLE ctas1 SELECT 'A' ")
+      sql(s"""CREATE TABLE ctas_with_existing_location stored as rcfile LOCATION
+           |'$tempLocation' AS SELECT key k, value FROM src ORDER BY k, value""".stripMargin)
+    }
+
+    Seq("false", "true").foreach { convertCTASFlag =>
+      Seq("false", "true").foreach { allowNonEmptyDirFlag =>
+        withSQLConf(
+          SQLConf.CONVERT_CTAS.key -> convertCTASFlag,
+          SQLConf.ALLOW_NON_EMPTY_LOCATION_IN_CTAS.key -> allowNonEmptyDirFlag) {
+          withTempDir { dir =>
+            val tempLocation = dir.toURI.toString
+            withTable("ctas1", "ctas_with_existing_location") {
+              if (!spark.conf.get(SQLConf.ALLOW_NON_EMPTY_LOCATION_IN_CTAS)) {
+                intercept[AnalysisException] {
```

Comment:
Please check the exact error message. Otherwise, we may miss some future regressions due to that another `AnalysisException`.

Reply:
Done

Reply:
@vinodkc hi, shall we intercept AnalysisException when `spark.sql.hive.convertCTAS=false` ?

Reply:
@PengleiShi , 
Could you please add a bit more details about this point.
I see `spark.sql.hive.convertCTAS` is deprecated from Spark 3.1  ([SPARK-33693](https://issues.apache.org/jira/browse/SPARK-33693)).

Reply:
> spark.sql.hive.convertCTAS

@insightslab Yes, i just mean `spark.sql.legacy.createHiveTableByDefault`,  i‘m confused about whether `spark.sql.legacy.allowNonEmptyLocationInCTAS` is worked when `spark.sql.legacy.createHiveTableByDefault=true`。It seems that CTAS hive table is not allowed with a exist dir even if `spark.sql.legacy.allowNonEmptyLocationInCTAS` is true.

Reply:
@PengleiShi , 
I did a test on the master branch
```
sql("set spark.sql.legacy.createHiveTableByDefault=true")
val tempLocation="/tmp/ctastest100"
sql(s"CREATE TABLE ctas1_h(id string) LOCATION '$tempLocation/ctas1'")
sql("INSERT INTO TABLE ctas1_h SELECT 'A' ")
sql(s"""CREATE TABLE ctas1_h_with_existing_location LOCATION '$tempLocation' AS SELECT 'B'""")
//Throws org.apache.spark.sql.AnalysisException: CREATE-TABLE-AS-SELECT cannot create table with location to a non-empty directory file...

sql("set spark.sql.legacy.allowNonEmptyLocationInCTAS=TRUE")
sql(s"""CREATE TABLE ctas1_h_with_existing_location LOCATION '$tempLocation' AS SELECT 'B'""".stripMargin)
// Above statement created table
```
Could please use this test to describe the expected behavior?

Reply:
I got it wrong. I missed the point  `with location`,  and always tested CTAS without a location and get unexpect result.   Now i understand,  many thx @vinodkc @insightslab